apps/src/music/views/MiniMusicPlayer.tsx

```diff
@@ -53,57 +62,70 @@ const MiniPlayerView: React.FunctionComponent<MiniPlayerViewProps> = ({
-      currentLibrary.setCurrentPackId(packId);
-      playerRef.current.updateConfiguration(
-        currentLibrary.getBPM(),
-        currentLibrary.getKey()
+  const onPlaySong = useCallback(
+    async (project: Channel) => {
+      playerRef.current.stopSong();
+      // Load code
+      const projectSources = await sourcesStoreRef.current.load(project.id);
+      workspaceRef.current.loadCode(
+        JSON.parse(projectSources.source as string)
       );
-    }
 
-    // Preload sounds in player
-    await playerRef.current.preloadSounds(
-      [...allTriggerEvents, ...sequencerRef.current.getPlaybackEvents()],
-      (loadTimeMs, soundsLoaded) => {
-        if (soundsLoaded > 0) {
-          Lab2Registry.getInstance()
-            .getMetricsReporter()
-            .reportLoadTime('MiniPlayer.SoundLoadTime', loadTimeMs);
-        }
-        Lab2Registry.getInstance().getMetricsReporter().logInfo({
-          event: 'MiniPlayerSoundsLoaded',
-          soundsLoaded,
-          loadTimeMs,
-          channelId: project.id,
-        });
+      // Compile song
+      workspaceRef.current.compileSong({Sequencer: sequencerRef.current});
+
+      // Execute compiled song
+      // Sequence out all possible trigger events to preload sounds if necessary.
+      sequencerRef.current.clear();
+      workspaceRef.current.executeAllTriggers();
+      const allTriggerEvents = sequencerRef.current.getPlaybackEvents();
+
+      sequencerRef.current.clear();
+      workspaceRef.current.executeCompiledSong();
+
+      // If there is a pack ID, give the player its BPM and key.
+      const currentLibrary = MusicLibrary.getInstance();
+      const packId = project.labConfig?.music.packId || null;
+      if (currentLibrary) {
+        currentLibrary.setCurrentPackId(packId);
+        playerRef.current.updateConfiguration(
+          currentLibrary.getBPM(),
+          currentLibrary.getKey()
+        );
       }
-    );
 
-    // Play sounds
-    playerRef.current.playSong(sequencerRef.current.getPlaybackEvents());
-    setCurrentProjectId(project.id);
-  }, []);
+      // Preload sounds in player
+      await playerRef.current.preloadSounds(
+        [...allTriggerEvents, ...sequencerRef.current.getPlaybackEvents()],
+        (loadTimeMs, soundsLoaded) => {
+          if (soundsLoaded > 0) {
+            Lab2Registry.getInstance()
+              .getMetricsReporter()
+              .reportLoadTime('MiniPlayer.SoundLoadTime', loadTimeMs);
+          }
+          Lab2Registry.getInstance().getMetricsReporter().logInfo({
+            event: 'MiniPlayerSoundsLoaded',
+            soundsLoaded,
+            loadTimeMs,
+            channelId: project.id,
+          });
+        }
+      );
+
+      // Play sounds
+      playerRef.current.playSong(sequencerRef.current.getPlaybackEvents());
+      setCurrentProjectId(project.id);
+
+      // Report analytics on play button.
+      analyticsReporter.onMiniMusicPlayerButtonClicked('play', {
```

Comment:
Curious why we're using a special `onMiniMusicPlayerButtonClicked` function here but then using the generic `onButtonClicked` function below? Would it make sense to always just use `onButtonClicked` and just pass in a specific button name like `"mini-player-play"` or similar?

Reply:
Oh! I meant for the one below to use `onMiniiMusicPlayerButtonClicked` but since i'm adding `sessionType`, I think it makes sense to use a mini-player-specific button name. Thanks!